### PR TITLE
cpu/kinetis: reworked PWM driver implementation

### PR DESCRIPTION
**Pull request by haukepetersen**
Some simplification and 'modern' style for the Kinetis PWM driver.

EDIT:
~~So far untested (hence WIP) and only adapted for the phyNODE and the mulle.~~

Confirmed working for the `pba-d-01-kw2x` (phyNODE)

### CODE REVIEW

**Comment by gebart** [comment]
I'll test it on mulle with a toy servo for a basic spot check.

**Comment by haukepetersen** [comment]
Thanks. I will test on the phyNODE later today

**Comment by haukepetersen** [comment]
Works fine on the phyNODE -> not WIP anymore.

File: cpu/kinetis_common/periph/pwm.c [hunk 1]
```diff
@@ -24,375 +24,108 @@
...
 
-        default:
-            return 0;
-    }
+uint32_t pwm_init(pwm_t pwm, pwm_mode_t mode, uint32_t freq, uint16_t res)
+{
+    uint8_t pre = 0;
 
-    if ((unsigned int)res > (PWM_MAX_VALUE + 1) || (res * freq) > pwm_clk) {
+    if (pwm >= PWM_NUMOF || ((res * freq) > CLOCK_BUSCLOCK)) {
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: cast `pwm` to `unsigned int` to avoid a Clang bug with enums

File: cpu/kinetis_common/periph/pwm.c [hunk 2]
```diff
@@ -24,375 +24,108 @@
...
     }
 
-    /* enable timer ergo the PWM generation */
-    pwm_start(dev);
+    /* and now we start the actual waveform generation */
+    pwm_start(pwm);
 
-    /* Return actual frequency */
-    return (pwm_clk / (1 << prescaler)) / res;
+    /* finally we need to return the actual applies PWM frequency */
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: applies->applied

File: cpu/kinetis_common/periph/pwm.c [hunk 3]
```diff
@@ -24,375 +24,108 @@
...
-    if (value > PWM_MAX_VALUE) {
-        value = PWM_MAX_VALUE;
-    }
-
-    /* cppcheck thinks ftmchan may be NULL here, but the variable is
-     * assigned in all non-returning branches of the switch at the top of
-     * this function. */
-    /* cppcheck-suppress nullPointer */
-    ftm->CONTROLS[ftmchan[channel]].CnV = value;
+    assert((pwm < PWM_NUMOF) && (channel < pwm_config[pwm].chan_numof));
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: cast to unsigned int

File: cpu/kinetis_common/periph/pwm.c [hunk 4]
```diff
@@ -24,375 +24,108 @@
...
-            return PWM_0_CHANNELS;
-#endif
-#if PWM_1_EN
-        case PWM_1:
-            return PWM_1_CHANNELS;
-#endif
-        default:
-            return 0;
-    }
+    assert(pwm < PWM_NUMOF);
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: cast to unsigned int

File: cpu/kinetis_common/periph/pwm.c [hunk 5]
```diff
@@ -24,375 +24,108 @@
...
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            PWM_1_DEV->SC |= FTM_SC_CLKS(1);
-            break;
-#endif
-    }
+    assert(pwm < PWM_NUMOF);
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: cast

File: cpu/kinetis_common/periph/pwm.c [hunk 6]
```diff
@@ -24,375 +24,108 @@
...
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            PWM_1_DEV->SC &= ~FTM_SC_CLKS_MASK;
-            break;
-#endif
-    }
+    assert(pwm < PWM_NUMOF);
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: cast

File: cpu/kinetis_common/periph/pwm.c [hunk 7]
```diff
@@ -24,375 +24,108 @@
...
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            PWM_1_CLKEN();
-            break;
-#endif
-    }
+    assert(pwm < PWM_NUMOF);
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: cast

File: cpu/kinetis_common/periph/pwm.c [hunk 8]
```diff
@@ -24,375 +24,108 @@
...
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            PWM_1_CLKDIS();
-            break;
-#endif
-    }
+    assert(pwm < PWM_NUMOF);
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: cast

File: cpu/kinetis_common/periph/pwm.c [hunk 9]
```diff
@@ -24,375 +24,108 @@
...
+    BITBAND_REG32(SIM->SCGC6, SIM_SCGC6_FTM0_SHIFT + pwm_config[pwm].ftm_num) = 0;
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: The FTM2 module clock enable bit is located in SCGC3, so this will only work for FTM0 and FTM1.

**Comment by haukepetersen** [comment]
@gebart thanks for the review, will fix everything requested.

File: cpu/kinetis_common/periph/pwm.c (hunk 1 above)
Comment (haukepetersen) [inline cpu/kinetis_common/periph/pwm.c]: IMHO that is unnecessary, as `pwm` is not an enum value (anymore). From pwm.h:
```c
#define PWM_DEV(x)          (x)
...
typedef unsigned int pwm_t;
```
So there should be no need for (ugly) casting here?!

**Comment by haukepetersen** [comment]
fixed doc typo and issue with `FTM2`.

@gebart: I tend not to fix the cast issue (see above). Could you verify?! Thanks

**Comment by gebart** [comment]
@haukepetersen I didn't realize that pwm_t had changed. I agree with not adding any casts there.

There's a problem building anything with clang right now, but it's not this PR's fault. I'll prepare a fix for it.

**Comment by haukepetersen** [comment]
Cool. So would you mind ACKing this PR then?

**Comment by haukepetersen** [comment]
squashed, waiting for final Murock run

File: cpu/kinetis_common/periph/pwm.c [hunk 10]
```diff
@@ -24,375 +24,120 @@
...
-            PWM_1_CLKEN();
-            break;
-#endif
+    assert(pwm < PWM_NUMOF);
+    int ftm = pwm_config[pwm].ftm_num;
+    if (ftm < 2) {
+        BITBAND_REG32(SIM->SCGC6, SIM_SCGC6_FTM0_SHIFT + ftm) = 1;
+    }
+    else if (ftm == 2) {
+        BITBAND_REG32(SIM->SCGC6, SIM_SCGC3_FTM2_SHIFT) = 1;
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: SIM->SCGC3

File: cpu/kinetis_common/periph/pwm.c [hunk 11]
```diff
@@ -24,375 +24,120 @@
...
-        case PWM_1:
-            PWM_1_CLKEN();
-            break;
-#endif
+    assert(pwm < PWM_NUMOF);
+    int ftm = pwm_config[pwm].ftm_num;
+    if (ftm < 2) {
+        BITBAND_REG32(SIM->SCGC6, SIM_SCGC6_FTM0_SHIFT + ftm) = 1;
+    }
+    else if (ftm == 2) {
```
Comment (gebart) [inline cpu/kinetis_common/periph/pwm.c]: Add a preprocessor `#if` that checks for the existence of the FTM2 module, it seems like the KW22 doesn't have it which resulted in build failures.

**Comment by haukepetersen** [comment]
@gebart Murdock is happy, do you re-ack?

**Comment by PeterKietzmann** [comment]
@gebart can you approve? I just re-tested this PR on a pba-d-01-kw2x and the output looks fine.

**Pull request merged**